fuel_test/prep_env.sh

```diff
@@ -0,0 +1,120 @@
+#!/bin/bash
+
+set +eux
+
+function restore_fm {
+	#Restore snapshot that has test machine's ssh public key
+	local xs_host="$1"
+	local fm_name="$2"
+	local fm_snapshot="$3"
+	ssh -qo StrictHostKeyChecking=no root@$xs_host \
+	'
+	set -eux
+	#echo "FM_NAME : '$fm_name'"
+	#echo "FM_SNAPSHOT : '$fm_snapshot'"
+	vm_uuid=$(xe vm-list name-label="'$fm_name'" --minimal)
+	#echo "VM_UUID : $vm_uuid"
+	snapshot_uuid=$(xe snapshot-list name-label="'$fm_snapshot'" snapshot-of="$vm_uuid" --minimal)
+	#echo "SNAPSHOT_UUID: $snapshot_uuid"
+	xe snapshot-revert snapshot-uuid="$snapshot_uuid"
+	xe vm-start vm="'$fm_name'"
+	'
+}
+
+function create_node {
+	# Create a node server(compute/controller node) with given memory, disk and NICs
+	local xs_host="$1"
+	local vm="$2"
+	local mem="$3"
+	local disk="$4"
+
+	ssh -qo StrictHostKeyChecking=no root@$xs_host \
+	'
+	set -eux
+
+	vm="'$vm'"
+	mem="'$mem'"
+	disk="'$disk'"
+
+	template="Other install media"
+
+	vm_uuid=$(xe vm-install template="$template" new-name-label="$vm")
+
+	localsr=$(xe pool-list params=default-SR --minimal)
+	extra_vdi=$(xe vdi-create \
+		name-label=xvdb \
+		virtual-size="${disk}GiB" \
+		sr-uuid=$localsr type=user)
+	vbd_uuid=$(xe vbd-create vm-uuid=$vm_uuid vdi-uuid=$extra_vdi device=0)
+	xe vm-cd-add vm=$vm_uuid device=1 cd-name="xs-tools.iso"
+
+	xe vm-memory-limits-set \
+		static-min=${mem}MiB \
+		static-max=${mem}MiB \
+		dynamic-min=${mem}MiB \
+		dynamic-max=${mem}MiB \
+		uuid=$vm_uuid
+
+	xe vm-param-set uuid=$vm_uuid HVM-boot-params:order=ndc
+	'
+}
+
+function add_network {
+	local xs_host="$1"
+	local vm="$2"
+	local network="$3"
+	local device="$4"
+	ssh -qo StrictHostKeyChecking=no root@$xs_host \
+	'
+	set -eux
+	vm="'$vm'"
+	network="'$network'"
+	device="'$device'"
+
+	vm_uuid=$(xe vm-list name-label="'$vm'" --minimal)
+	network=${network//Network /Pool-wide network associated with eth}
+	network_uuid=$(xe network-list name-label="$network" --minimal)
+	xe vif-create network-uuid=$network_uuid vm-uuid=$vm_uuid device=$device
+	'
+}
+
+function add_himn {
+	# Add HIMN to given compute node and return the Mac address of the added NIC
+	local xs_host="$1"
+	local vm="$2"
+	ssh -qo StrictHostKeyChecking=no root@$xs_host \
+	'
+	set -eux
+	vm="'$vm'"
+	network=$(xe network-list bridge=xenapi minimal=true)
+	vm_uuid=$(xe vm-list name-label="'$vm'" --minimal)
+
+	vif=$(xe vif-list network-uuid=$network vm-uuid=$vm_uuid --minimal)
+	if [ -z $vif ]; then
+		vif=$(xe vif-create network-uuid=$network vm-uuid=$vm_uuid device=9 minimal=true)
+	fi
+
+	xe vm-start vm=$vm
+	mac=$(xe vif-list uuid=$vif params=MAC --minimal)
```

Comment:
I started compute node earlier because I found if the vm hasn't been started, the dom-id will always be -1 so I cannot add himn to xenstore. In SDK it is more straightforward because vm-uuid instead of dom-id is required.